File: androidApp/src/main/java/me/nathanfallet/bdeensisa/features/clubs/ClubCard.kt
package me.nathanfallet.bdeensisa.features.clubs

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.material.Card
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import me.nathanfallet.bdeensisa.models.Club

@Composable
fun ClubCard(
    club: Club,
    badgeText: String?,
    badgeColor: Color,
    action: (() -> Unit)?,
    detailsEnabled: Boolean,
    showDetails: ((Club) -> Unit)? = null
) {

    Card(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp)
            .padding(vertical = 4.dp)
            .clickable {
                if (detailsEnabled) {
                    showDetails?.invoke(club)
                }
            },
        elevation = 4.dp
    ) {
        Column(
            modifier = Modifier
                .padding(16.dp)
                .fillMaxWidth()
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                Column(
                    modifier = Modifier.weight(1f, fill = false)
                ) {
                    Text(
                        text = club.name,
                        fontWeight = FontWeight.Bold
                    )
                    Text("${club.membersCount} membre${if (club.membersCount != 1L) "s" else ""}")
                }
                badgeText?.let {
                    Text(
                        text = badgeText,
                        style = MaterialTheme.typography.caption,
                        color = Color.White,
                        modifier = Modifier
                            .background(badgeColor, MaterialTheme.shapes.small)
                            .padding(horizontal = 10.dp, vertical = 6.dp)
                            .clickable {
                                action?.invoke()
                            }
                    )
                }
            }
            Spacer(modifier = Modifier.height(8.dp))
            Text(club.description ?: "")
        }
    }

}

File: iosApp/iosApp/Features/Clubs/ClubCard.swift
//
//  ClubCard.swift
//  BDE
//
//  Created by Nathan Fallet on 13/02/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import SwiftUI
import Kingfisher
import shared

struct ClubCard: View {
    
    let club: Club
    let badgeText: String?
    let badgeColor: Color?
    let action: ((Club) -> Void)?
    let detailsEnabled: Bool
    
    @State var showDetails = false
    
    var body: some View {
        NavigationLink(
            destination: ClubView(viewModel: ClubViewModel(club: club)),
            isActive: $showDetails
        ) {
            VStack(alignment: .leading, spacing: 8) {
                HStack {
                    if let logo = club.logo, let url = URL(string: "https://bdensisa.org/clubs/\(club.id)/uploads/\(logo)") {
                        KFImage(url)
                            .resizable()
                            .frame(maxWidth: 44, maxHeight: 44)
                            .cornerRadius(4)
                    }
                    VStack(alignment: .leading) {
                        Text(club.name)
                            .fontWeight(.bold)
                        Text("\(club.membersCount ?? 0) membre\(club.membersCount ?? 0 != 1 ? "s" : "")")
                    }
                    Spacer()
                    if let badgeText, let badgeColor {
                        if let action {
                            Button(badgeText) {
                                action(club)
                            }
                            .font(.caption)
                            .padding(.horizontal, 10)
                            .padding(.vertical, 6)
                            .foregroundColor(.white)
                            .background(badgeColor)
                            .cornerRadius(8)
                        } else {
                            Text(badgeText)
                                .font(.caption)
                                .padding(.horizontal, 10)
                                .padding(.vertical, 6)
                                .foregroundColor(.white)
                                .background(badgeColor)
                                .cornerRadius(8)
                        }
                    }
                }
                Text(club.description_ ?? "")
                    .lineLimit(detailsEnabled ? 5 : nil)
            }
            .foregroundColor(.primary)
            .multilineTextAlignment(.leading)
            .cardView()
            .onTapGesture {
                if detailsEnabled {
                    showDetails = true
                }
            }
        }
    }
    
}
